Init project

forhadmethun opened the pull request.

forhadmethun committed:
Init project

grzegorz-bielski commented on src/main/scala/models/Messages.scala:
```diff
@@ -0,0 +1,18 @@
+  final case class InitMessageResponse(
+      `type`: String,
+      data: InitMessageResponseData
+  ) derives Codec.AsObject
```
```scala
enum Response derives Codec.Enum:
  case Init(data: InitMessageResponseData)
  case Sth
```

or
```scala
trait Response

final case class Init() extends Response
```

```scala
trait Request // or. 
enum Request
```

See https://github.com/circe/circe/pull/1800 for Circe Scala 3 derivation

forhadmethun replied:
Segregated Request and Response messages with sealed trait 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/models/Messages.scala:
```diff
@@ -0,0 +1,18 @@
+  final case class InitMessageResponseData(orbs: List[Orb])
+      derives Codec.AsObject
+  final case class InitMessageResponse(
+      `type`: String,
```
You can use `Configuration.default.withDiscriminator`

forhadmethun replied:
Renamed the variable with `messageType` to keep it simpler
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on build.sbt:
```diff
@@ -0,0 +1,23 @@
+      "org.http4s" %% "http4s-circe" % Http4sVersion,
+      "org.http4s" %% "http4s-dsl" % Http4sVersion,
+      "io.circe" %% "circe-core" % CirceVersion,
+      "io.circe" %% "circe-generic" % CirceVersion,
```
You might not need it

forhadmethun replied:
It's actually necessary to get a generic decoder for the following portion https://github.com/forhadmethun/agar-winner/blob/f6ebfd8edf9498933921e6c56dae6e010a08d839/src/main/scala/services/GameService.scala#L41

grzegorz-bielski commented on src/main/scala/models/Orb.scala:
```diff
@@ -0,0 +1,28 @@
+    val locX = scala.util.Random.nextInt(worldWidth)
+    val locY = scala.util.Random.nextInt(worldHeight)
+    val radius = 5
+    Orb(color, locX, locY, radius)
```
Use IO for side effect such as Random

forhadmethun replied:
Updated the Orb.scala with effect :) 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/models/Orb.scala:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  def generateOrbs(settings: Settings): List[Orb] = {
+    List.fill(settings.defaultOrbs)(
```
Try changing to Vector or ArraySeq

forhadmethun replied:
Updated with Vector 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/models/Player.scala:
```diff
@@ -0,0 +1,22 @@
+    val g = scala.util.Random.nextInt(150) + 50
+    val b = scala.util.Random.nextInt(150) + 50
+    s"rgb($r,$g,$b)"
+  }
```
```scala
object PlayerData:
  def createPlayer(...): IO[PlayerData]
 ```

forhadmethun replied:
Updated the Player.scala with effect
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/models/Player.scala:
```diff
@@ -0,0 +1,22 @@
+  val radius: Int = settings.defaultSize
+  val color: String = getRandomColor()
+  var score: Int = 0
+  var orbsAbsorbed: Int = 0
```
```scala
final case class PlayerData(score: Int, ....): 
  def increaseScore: PlayerData = copy(
       score = score + 1
   )
```

forhadmethun replied:
Added method for such case
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/models/Settings.scala:
```diff
@@ -0,0 +1,10 @@
+    defaultZoom: Double = 1.5,
+    worldWidth: Int = 1440,
+    worldHeight: Int = 1440
+)
```
Later try `pureconfig` or `Ciris`

forhadmethun replied:
Will do in later phase :)

grzegorz-bielski commented on src/main/scala/routes/GameRoutes.scala:
```diff
@@ -0,0 +1,35 @@
+      })
+
+      val in: Pipe[F, WebSocketFrame, Unit] =
+        _.evalMap(a => Sync[F].delay(scribe.info(a.toString)).as(a))
```
You can check https://typelevel.org/log4cats/ or https://github.com/evolution-gaming/cats-helper

forhadmethun replied:
Added log4cats
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/routes/GameRoutes.scala:
```diff
@@ -0,0 +1,35 @@
+      })
+
+      val in: Pipe[F, WebSocketFrame, Unit] =
+        _.evalMap(a => Sync[F].delay(scribe.info(a.toString)).as(a))
```
`evalTap`

forhadmethun replied:
Updated with evalTap
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/services/GameService.scala:
```diff
@@ -0,0 +1,91 @@
+        Spawn[F].start(game.deamon).void
+      })
+    }
+    Spawn[F].start(next).void
```
You could also try
```
Spawn[F].background
```

forhadmethun replied:
For now, I'm not changing it, so moving forward with the start method but with single fiber
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

grzegorz-bielski commented on src/main/scala/services/GameService.scala:
```diff
@@ -0,0 +1,91 @@
+      topic <- Topic[F, GameMessage]
+    yield Game(topic, q)
+
+final case class GameMessage(id: String, content: String) derives Codec.AsObject
```
You don't codec here, probably

forhadmethun replied:
Removed this unnecessary codec
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

forhadmethun committed:
Fix feedbacks
- Segregate Request and Response messages
- Change `type` var to `messageType`
- IO for side effects (Random, UUID) in Orb.scala, Player.scala
- Use Vector instead of List
- Use evalTap instead of evalMap
- Remove unnecessary codec
- Use log4cats for logging
- Functional methods for increasing score / orbsAbsorbed count

forhadmethun committed:
Fix feedbacks
- Fix `publish` method og GameService to take GameMessage instead of F[GameMessage]
